### ISSUE TITLE
Blocking Timeout has different behavior in HttpInput vs HttpOutput

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#1958]
the blocking timeout was originally created to apply to individual IO operations.  However in HttpInput this has been changed to apply to the entire stream.  This change needs to be reviewed and the behaviour made consistent and better documented.

**Comment by gregw** [comment]
@sbordet I'm thinking this may be best fixed in HttpOutput.

On HttpInput, if the blocking timeout is applied to an individual read operation, then it becomes no different to the idle timeout.  On doing a blocking read, if there is data available it will be immediately returned.  If there is not data available, it will block waiting for content and during this time, both timeouts apply.  If any data arrives, the read is no longer idle and returns immediately with that data, also ending the blocking timeout.

On HttpOutput, the blocking timeout is currently applied to just an individual write.  It is different to the idle timeout, but is difficult to use because the size of the buffer greatly influences what the timeout means.    

So perhaps we should make the same change to HttpOutput, ie so the blockingTimeout applies to the total stream, over multiple writes?

**Comment by gregw** [comment]
Note also we have a suggestion that the blockingTimeout for HttpInput should be implemented as a timeout budget.     Currently it is set for an absolute time from the first read operation, but this may include processing time that is not part of the blocking.  For example if the timeout is set to 5000 the following code may timeout even if data is sent:
```java
int b0 = in.read();
Thread.sleep(10_000);
int b1 = in.read()
```

The suggestion is that we should keep a timeout budget and the time spent in each write is subtracted from the budget and used as the next read operation timeout.

**Comment by gregw** [comment]
This is an update on the status of this issue and PRs, inviting feedback from users to help decide the future direction of blocking timeout.

Firstly some background.   The blocking timeout was first introduced perversely as we were implementing support of asynchronous IO.   The issue was that our early async implementations had some bugs and occasionally a callback would be lost.  However, as we had changed our blocking implementation to just be a facade over async, a lost callback would result in a thread blocked forever and eventually server death by thread starvation.     The blocking timeout was implemented as a safety net as it was better to incorrectly timeout a broken IO operation rather than block forever.   The important thing to note is that the timeout was never intended to be a primary user configuration, rather a defensive implementation detail.

However, the timeout started to be used by users as part of their protection against slow clients and thus there was demand to make it more usable.   But the problem with a blocking timeout for DOS protection is that it's semantic is entirely dependent on buffer size and is also significantly different for read side (where partial buffer can be returned) and write side (where complete buffer must be flushed).

#845 partially addressed this by adding a read side data rate limit, which essentially abstracted away from the buffer size issue.   However it did change the meaning of the blocking timeout on the read side so it now applies as a total time limit from the first read to the last.     This is significantly different to the write side and as it includes processing time, may not be exactly the right semantic.

PR #1961 has been proposed as a way to make the blocking timeout more consistent and understandable for both reads and writes.     It converts the blocking timeout to a total blocking budget over multiple operations.  Essentially it sets a total time that can be spent blocking - either on 1 or many operations, before a timeout occurs.  I think this is a good change as it makes the blocking timeout useful as an upper bound on how long request handling can block.   However, it is a behaviour change that we need to consider if it should go into 9.4.x or be punted to 10.0.0

Separately, even the updated blocking timeout has a limitation in that it is not reactive to total size of the content.  The same blocking timeout is applied to 16 byte request body as to a 4GB file upload.   Thus if your application has large differences in content size, the upper bound for a blocking timeout may need to be large.  Also "how long spent blocking" during a 4GB file upload may actually be shorter than for the 16 byte body if the upload is providing a steady stream of data faster than the server consumes it.

For the file upload, we already have implemented a request data rate limit in #845, thus it is probably sensible to also implement a response data rate limit for downloading 4GB files.  As this would be a new feature, it can be added to 9.4.x even if we decide to delay #1961 to Jetty-10

tl;dr; We need a PR to add a response data rate; We need feedback to decide if #1961 should be merged to 9.4.x or 10.0.0

**Comment by sbordet** [comment]
I think we should remove completely blockingTimeout in Jetty 10, and just keep idle timeout and data rate on request and on response.

Perhaps the current effort of a better implementation for blockingTimeout can go into 9.4.x (and then removed in 10.0.x).
If people moving from 9.4.x to 10.0.x complains about the lack of this feature, would be easy to forward-port it.

**Comment by gregw** [comment]
I tend to agree, but would advocate a slower transition:
+ fix it in 9.4.x
+ deprecate it in 10.0.x so that it warns if used.
+ remove it in 10.1.x

I don't want to remove it only for a user to come up with a really good use-case and then struggle to get it back in.

**Comment by gregw** [comment]
We have decided to leave the blocking timeout as is in 9.4.x but instead implement the minResponseRate . Thus #1961 has been closed for now, but may be reactivated later for jetty-10.

**Comment by gregw** [comment]
See #2012

**Comment by gregw** [comment]
replaced by #2525
